### PR TITLE
Fixes #29628 - add action cable with redux channel

### PR DESCRIPTION
**Pull request by Ron-Lavi**
Setup action cable with a redux channel,
it means that we could dispatch redux actions from the server which will update the client-side redux store !

by that we could achieve many things, for example:
- stop doing API polling that often returns no updates from the server.
- show toast notifications without reloading the page
- use loading bars

and more..

as an example I removed the notification polling WIP commit,
and instead dispatching some redux actions from the server

### CODE REVIEW

File: config/cable.yml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+development:
+  adapter: async
+
+test:
+  adapter: async
+
+production:
+  adapter: async
```
Comment (Ron-Lavi) [inline config/cable.yml:8]: what should we use in production?

**Comment by ares** [comment]
I'm really happy to see this effort! cc @amirfefer, as we discussed last week, this could be useful on host detail page

**Comment by timogoebel** [comment]
> this could be useful on host detail page

It would be great to see the host statuses update on the fly. :-)


I'm wondering if we need anycable or a `websocket proxy` definition for this to work in production.

**Comment by ohadlevy** [comment]
how do we connect websockets to apache ? I would prefer to avoid another
port if we can?

File: config/cable.yml (hunk 1 above)
Comment (Ron-Lavi) [reply config/cable.yml:8]: on RoR guides they mention Redis, is it something we support? 
checkout: https://edgeguides.rubyonrails.org/action_cable_overview.html#subscription-adapter

```yml
production:
  adapter: redis
  url: redis://10.10.3.153:6381
  channel_prefix: appname_production
```

**Comment by ekohl** [comment]
> how do we connect websockets to apache ? I would prefer to avoid another port if we can?

Our Puma deployment supports that already, as verified by a community member who uses it in his own plugin.

It does mean we have to drop Passenger support once we start to rely on this. With Foreman 2.1 we're making Puma the default and migrate users over. Depending on how that goes, we'll decide on when we can drop Passenger. Best case scenario there are no issues and 2.2 will drop it.

File: config/cable.yml (hunk 1 above)
Comment (ekohl) [reply config/cable.yml:8]: Due to Dynflow there is a Redis instance present and configured for persistence so either that or PostgreSQL can be counted on I think. I don't know how the two compare.

**Comment by ekohl** [comment]
I'm setting this to draft since IMHO we can't merge this until we drop Passenger, which won't happen within a few weeks.

**Comment by ares** [comment]
Is this still draft? I think passenger is gone now or not?

**Comment by ekohl** [comment]
> Is this still draft? I think passenger is gone now or not?

There has been no decision on when to drop Passenger. Right now it's the new default in 2.1 but that was only changing the value for `--foreman-passenger` from `true` to `false`. https://projects.theforeman.org/issues/29777 tracks removal. Delaying this gives us the chance to roll back in case there are issues. My personal preference is to only drop it 2 or 3 weeks after 2.1 has gone GA with Puma. That's when you know it really works for the majority of users.

**Comment by ares** [comment]
ok, makes sense, I though we'd drop it earlier

File: app/controllers/api/v2/base_controller.rb [hunk 2]
```diff
@@ -165,6 +165,10 @@ def disable_json_root
         # re-enable json root element
         ActiveRecord::Base.include_root_in_json = true
       end
+
+      def redux_action(action)
+        ActionCable.server.broadcast 'redux_channel', action: action
```
Comment (sharvit) [inline app/controllers/api/v2/base_controller.rb:170]: ATM your are allowing the server the dispatch any action in the client by the stream. It's a nice POC but a bad practice in API communication.

IMO the right approach should be a dedicated stream for each action, for example:
```rb
# in host creation
ActionCable.server.broadcast 'Hosts', host: @host
# in host update
ActionCable.server.broadcast 'Hosts#1234', host: @host
```

Or a main stream, but deliver the message to the right subscriber:
```rb
# in host creation
ActionCable.server.broadcast 'web_notifications', model: 'Host', action: 'create', host: @host
# in host update
ActionCable.server.broadcast 'web_notifications', model: 'Host', action: 'update', host: @host
```

I'm not sure what’s right production wise but IMO it would be easier and cleaner to implement the first one.

Comment (Ron-Lavi) [reply app/controllers/api/v2/base_controller.rb:170]: With the `redux_channel` implementation, I tried to keep it generic as possible so the action will determine the scope of the component that will get affected.

What do you think is the benefit of having multiple channels? In my vision I think that every channel broadcast should trigger a redux store update, as this is our client architecture, that's why I combine it into one channel.

I think that in each implementation of it, if it's not something like notifications.. we should let the server know if that component is visible to prevent unnecessary broadcasts, maybe than we should handle a queue for the time the component is visible again, does `channels` handle that for us? or just broadcast whatever they get?

File: webpack/assets/javascripts/channels/redux_channel.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+  connected() {
+    // Called when the subscription is ready for use on the server
+  },
+
+  disconnected() {
+    // Called when the subscription has been terminated by the server
+  },
+
+  received({ action }) {
+    store.dispatch(action);
```
Comment (amirfefer) [inline webpack/assets/javascripts/channels/redux_channel.js:14]: IMHO this approach is somewhat limited and confusing, the server responds only with actions
instead, a client action can just subscribe to a channel, and store its data in a unique key in our store, just like our API ( a middleware can fit for this).

**Review by amirfefer — COMMENTED** [review]
Thanks @laviro ! awsome implementation! :+1: 
I have only one concern with the "server-action" design

File: app/controllers/api/v2/base_controller.rb (hunk 2 above)
Comment (sharvit) [reply app/controllers/api/v2/base_controller.rb:170]: > What do you think is the benefit of having multiple channels? In my vision I think that every channel broadcast should trigger a redux store update, as this is our client architecture, that's why I combine it into one channel.

The advantages are:

1. It would not cause a spaghetti code as we have with `jquery`
2. Easier to debug
3. Instead of the server will send all the messages to all the subscribers it will send specific messages to the `user/organization/page/chrome-tab`.
4. The server will stay responsible to bring you the data you need while the client will continue to react based on the data.
5. It won't be coupled to the client-side code, which means we can use it in other tools like hammer or any other external UI tool the community would wish to create.

> I think that in each implementation of it, if it's not something like notifications.. we should let the server know if that component is visible to prevent unnecessary broadcasts, maybe than we should handle a queue for the time the component is visible again, does channels handle that for us? or just broadcast whatever they get?


Multiple channels solve that by bringing the choice to the client so it can decide what channels to subscribe to.

In other words, if I'm on the show host page and the id is `1234`, I would subscribe to the `Hosts#1234` channel and unsubscribe once I leave the page.

**Comment by ezr-ondrej** [comment]
I believe we can revisit this, because we don't support passanger now.

**Comment by ekohl** [comment]
Linking back the RFC to drop Passenger: https://community.theforeman.org/t/drop-passenger-support/22787

**Comment by Ron-Lavi** [comment]
Awesome work guys! ok, I will get back to work on this soon

**Comment by Ron-Lavi** [comment]
I was hoping to get into it sooner, closing for now and I guess I will reopen and start fresh as many things changed
